src/libs/safe-path/src/scoped_dir_builder.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) 2022 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::io::{Error, ErrorKind, Result};
+use std::path::{Path, PathBuf};
+
+use crate::{scoped_join, PinnedPathBuf};
+
+const DIRECTORY_MODE_DEFAULT: u32 = 0o700;
+const DIRECTORY_MODE_MASK: u32 = 0o777;
+
+/// Safe version of `DirBuilder` to protect from TOCTOU style of attacks.
+///
+/// The `ScopedDirBuilder` is a counterpart for `DirBuilder`, with safety enhancements of:
+/// - ensuring the new directories are created under a specified `root` directory.
+/// - avoiding symlink race attacks during making directories.
+/// - returning a [PinnedPathBuf] for the last level of directory, so it could be used for other
+///   operations safely.
+#[derive(Debug)]
+pub struct ScopedDirBuilder {
+    root: PinnedPathBuf,
+    mode: u32,
+    recursive: bool,
+}
+
+impl ScopedDirBuilder {
+    /// Create a new instance of `ScopedDirBuilder` with with default mode/security settings.
+    pub fn new<P: AsRef<Path>>(root: P) -> Result<Self> {
+        let root = PinnedPathBuf::from_path(root)?;
+        if !root.is_dir() {
+            return Err(Error::new(
+                ErrorKind::Other,
+                format!("Invalid root path: {}", root.display()),
+            ));
+        }
+
+        Ok(ScopedDirBuilder {
+            root,
+            mode: DIRECTORY_MODE_DEFAULT,
+            recursive: false,
+        })
+    }
+
+    /// Indicates that directories should be created recursively, creating all parent directories.
+    ///
+    /// Parents that do not exist are created with the same security and permissions settings.
+    pub fn recursive(&mut self, recursive: bool) -> &mut Self {
+        self.recursive = recursive;
+        self
+    }
+
+    /// Sets the mode to create new directories with. This option defaults to 0o755.
+    pub fn mode(&mut self, mode: u32) -> &mut Self {
+        self.mode = mode & DIRECTORY_MODE_MASK;
+        self
+    }
+
+    /// Creates the specified directory with the options configured in this builder.
+    ///
+    /// The `path` must be a subdirectory of `ScopedDirBuilder::root`, otherwise error will be
+    /// returned. It is considered an error if the directory already exists unless recursive mode
+    /// is enabled.
+    pub fn create<P: AsRef<Path>>(&self, path: P) -> Result<PinnedPathBuf> {
+        let path = scoped_join("/", path)?;
+        self.do_mkdir(&path)
+    }
```

Comment:
Thanks for the suggestion, renamed `create` to `create_with_absolute_path`, and also renamed `create_sub_directory` to simply `create` with refined comments.

Reply:
Excellent.  I think we can still improve the name `create_with_absolute_path` though: `create` can also take an absolute path, but it will be interpreted within `root` (as though it were a chroot or namespace).  So maybe `create_with_unscoped_path` for the other one?